### ISSUE TITLE
Do not Goofys support AWS Signature Version 4?

### ISSUE THREAD

**Issue by go201812** [kahing/goofys#386]
I can mount S3 bucket using goofys, but can not upload/download the file from S3 bucket.
I can upload/download from S3 bucket by AWS CLI using same credential and config.

My Enviroment is below.
■OS
# uname -r
3.10.0-514.21.1.el7.x86_64

■MW
# goofys --version
goofys version 0.19.0-943e017724ea820eb4185419ef3c41d6f921a324

# go version
go version go1.9.4 linux/amd64

■AWS S3
S3 Bugets is encrypted by SSE-KMS.

Executed Command:
・Execute S3 Mount Command:
  /usr/local/bin/goofys XXXX-XXXX /mnt/mountpoint -o nonempty,allow_other,--endpoint=http://s3.ap-northeast-1.amazonaws.com/, --sse-kms=xxx-xxx-xxx-xxx,--debug_s3,--debug_fuse,-f

・Download File Command:
  cp /mnt/mountpoint/ca/test11.txt /tmp/fromS3_test.txt

Error Message is below.
"Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4." is occured

[Tue Dec 18 10:35:03.282 2018] 2018/12/18 10:35:03.242034 fuse.DEBUG ReadFile 21 ca/test11.txt [0 4096]
[Tue Dec 18 10:35:03.282 2018] 2018/12/18 10:35:03.242350 s3.DEBUG DEBUG: Request s3/GetObject Details:
[Tue Dec 18 10:35:03.282 2018] ---[ REQUEST POST-SIGN ]-----------------------------
[Tue Dec 18 10:35:03.282 2018] GET /XXXX-XXXX/ca/test11.txt HTTP/1.1
[Tue Dec 18 10:35:03.282 2018] Host: s3.ap-northeast-1.amazonaws.com
[Tue Dec 18 10:35:03.282 2018] User-Agent: aws-sdk-go/1.8.25 (go1.9.2; linux; amd64)
[Tue Dec 18 10:35:03.282 2018] Accept-Encoding: identity
[Tue Dec 18 10:35:03.282 2018] Authorization: AWS XXXX:XXXX
[Tue Dec 18 10:35:03.282 2018] X-Amz-Date: Tue, 18 Dec 2018 01:35:03 +0000
[Tue Dec 18 10:35:03.282 2018] 
[Tue Dec 18 10:35:03.282 2018] 
[Tue Dec 18 10:35:03.282 2018] -----------------------------------------------------
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.329978 s3.DEBUG DEBUG: Response s3/GetObject Details:
[Tue Dec 18 10:35:03.329 2018] ---[ RESPONSE ]--------------------------------------
[Tue Dec 18 10:35:03.329 2018] HTTP/1.1 400 Bad Request
[Tue Dec 18 10:35:03.329 2018] Transfer-Encoding: chunked
[Tue Dec 18 10:35:03.329 2018] Connection: keep-alive
[Tue Dec 18 10:35:03.329 2018] Content-Type: application/xml
[Tue Dec 18 10:35:03.329 2018] Date: Tue, 18 Dec 2018 01:35:03 GMT
[Tue Dec 18 10:35:03.329 2018] Server: nginx/1.12.1
[Tue Dec 18 10:35:03.329 2018] X-Amz-Id-2: xxxxx
[Tue Dec 18 10:35:03.329 2018] X-Amz-Region: ap-northeast-1
[Tue Dec 18 10:35:03.329 2018] X-Amz-Request-Id: 055EF5D5D60CD708
[Tue Dec 18 10:35:03.329 2018] 
[Tue Dec 18 10:35:03.329 2018] 
[Tue Dec 18 10:35:03.329 2018] -----------------------------------------------------
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330012 s3.DEBUG DEBUG: Validate Response s3/GetObject failed, not retrying, error InvalidArgument: Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.
[Tue Dec 18 10:35:03.329 2018] 	status code: 400, request id: 055EF5D5D60CD708, host id: xxxxx
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330029 fuse.DEBUG < readFromStream 21 ca/test11.txt [0]
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330051 fuse.DEBUG < readFile 21 ca/test11.txt [0 invalid argument]
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330080 fuse.DEBUG < ReadFile 21 ca/test11.txt [0 invalid argument]
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330106 fuse.DEBUG Op 0x0000003d        connection.go:481] -> Error: "invalid argument"
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330243 fuse.DEBUG Op 0x0000003e        connection.go:396] <- ReadFile (inode 21, handle 3, offset 0, 4096 bytes)
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330511 fuse.DEBUG ReadFile 21 ca/test11.txt [0 4096]
[Tue Dec 18 10:35:03.329 2018] 2018/12/18 10:35:03.330753 s3.DEBUG DEBUG: Request s3/GetObject Details:
[Tue Dec 18 10:35:03.329 2018] ---[ REQUEST POST-SIGN ]-----------------------------
[Tue Dec 18 10:35:03.329 2018] GET /XXXX-XXXX/ca/test11.txt HTTP/1.1
[Tue Dec 18 10:35:03.329 2018] Host: s3.ap-northeast-1.amazonaws.com
[Tue Dec 18 10:35:03.329 2018] User-Agent: aws-sdk-go/1.8.25 (go1.9.2; linux; amd64)
[Tue Dec 18 10:35:03.329 2018] Accept-Encoding: identity
[Tue Dec 18 10:35:03.329 2018] Authorization: AWS xxxx
[Tue Dec 18 10:35:03.329 2018] X-Amz-Date: Tue, 18 Dec 2018 01:35:03 +0000
[Tue Dec 18 10:35:03.329 2018] 
[Tue Dec 18 10:35:03.329 2018] 
[Tue Dec 18 10:35:03.329 2018] -----------------------------------------------------
[Tue Dec 18 10:35:03.413 2018] 2018/12/18 10:35:03.397137 s3.DEBUG DEBUG: Response s3/GetObject Details:
[Tue Dec 18 10:35:03.413 2018] ---[ RESPONSE ]--------------------------------------
[Tue Dec 18 10:35:03.413 2018] HTTP/1.1 400 Bad Request
[Tue Dec 18 10:35:03.413 2018] Transfer-Encoding: chunked
[Tue Dec 18 10:35:03.413 2018] Connection: keep-alive
[Tue Dec 18 10:35:03.413 2018] Content-Type: application/xml
[Tue Dec 18 10:35:03.413 2018] Date: Tue, 18 Dec 2018 01:35:03 GMT
[Tue Dec 18 10:35:03.413 2018] Server: nginx/1.12.1
[Tue Dec 18 10:35:03.413 2018] X-Amz-Id-2: xxxx
[Tue Dec 18 10:35:03.413 2018] X-Amz-Region: ap-northeast-1
[Tue Dec 18 10:35:03.413 2018] X-Amz-Request-Id: 9F4E4456ED7BCD66
[Tue Dec 18 10:35:03.413 2018] 
[Tue Dec 18 10:35:03.413 2018] 
[Tue Dec 18 10:35:03.413 2018] -----------------------------------------------------
[Tue Dec 18 10:35:03.413 2018] 2018/12/18 10:35:03.397176 s3.DEBUG DEBUG: Validate Response s3/GetObject failed, not retrying, error InvalidArgument: Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.
[Tue Dec 18 10:35:03.413 2018] 	status code: 400, request id: 9F4E4456ED7BCD66, host id: xxxxx
[Tue Dec 18 10:35:03.413 2018] 2018/12/18 10:35:03.397188 fuse.DEBUG < readFromStream 21 ca/test11.txt [0]
[Tue Dec 18 10:35:03.413 2018] 2018/12/18 10:35:03.397196 fuse.DEBUG < readFile 21 ca/test11.txt [0 invalid argument]
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.397203 fuse.DEBUG < ReadFile 21 ca/test11.txt [0 invalid argument]
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.397235 fuse.DEBUG Op 0x0000003e        connection.go:481] -> Error: "invalid argument"
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.397990 fuse.DEBUG Op 0x0000003f        connection.go:396] <- FlushFile (inode 21)
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398122 fuse.DEBUG FlushFile 21 ca/test11.txt []
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398136 fuse.DEBUG <-- FlushFile 21 ca/test11.txt [<nil>]
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398154 fuse.DEBUG Op 0x0000003f        connection.go:479] -> OK ()
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398245 fuse.DEBUG Op 0x00000040        connection.go:396] <- ReleaseFileHandle
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398451 fuse.DEBUG ReleaseFileHandle ca/test11.txt
[Tue Dec 18 10:35:03.429 2018] 2018/12/18 10:35:03.398484 fuse.DEBUG Op 0x00000040        connection.go:479] -> OK ()

**Comment by t2y** [comment]
@kahing AWS announced Signature Version 2 will end on soon. I think this issue's priority is high. Do you have a plan to support Signature Version 4?

> Signature Version 2 is being deprecated, and the final support for Signature Version 2 will end on June 24, 2019. After June 24, 2019, Amazon S3 will only accept API requests that are signed using Signature Version 4.

https://docs.aws.amazon.com/AmazonS3/latest/dev/auth-request-sig-v2.html

**Comment by gaul** [comment]
goofys should support v4 signatures via the AWS SDK.  Can you share the steps to demonstrate that it does not?

**Comment by t2y** [comment]
hmm, in my environment, goofys tried to request with v4 signature, but it failed. then, falling back to v2 signature.

```
2019/03/11 14:08:00.680559 s3.DEBUG DEBUG: Request s3/HeadObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
HEAD /348l6878talfyi7u19vs8deysa0ej80v HTTP/1.1
Host: ********.s3-ap-northeast-1.amazonaws.com
User-Agent: aws-sdk-go/1.8.25 (go1.12; darwin; amd64)
Authorization: AWS4-HMAC-SHA256 Credential=********/20190311/s3-ap-northeast-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=********57f5edc1e7100bac85721336036ce21009bac07950c582c0
X-Amz-Content-Sha256: ********1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20190311T050800Z


-----------------------------------------------------
2019/03/11 14:08:00.998530 s3.DEBUG DEBUG: Response s3/HeadObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Mon, 11 Mar 2019 05:08:00 GMT
Server: AmazonS3
X-Amz-Id-2: ********F1BRqQkKmbTRvSCNCfGpULvS6DcAfFXKdSwehsFKKZ6y95C5//5xW01qoWsOc=
X-Amz-Request-Id: ********D95ABEFCE



-----------------------------------------------------
2019/03/11 14:08:00.998647 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error BadRequest: Bad Request
    status code: 400, request id: B524017D95ABEFCE, host id: ********UtadF1BRqQkKmbTRvSCNCfGpULvS6DcAfFXKdSwehsFKKZ6y95C5//5xW01qoWsOc=
2019/03/11 14:08:00.998702 s3.INFO Falling back to v2 signer
```

**Comment by gaul** [comment]
Can you try with some other regions and report their results, including whether v2 or v4 signatures were negotiated?

**Comment by t2y** [comment]
> Can you try with some other regions and report their results, including whether v2 or v4 signatures were negotiated?

Currently, I don't have some other regions, but I can create and confirm.

I run goofys with as below command, and above log is first request log in the debug log. I guess first head-object request has failed, so goofys did fallback to v2 signature.

{code}
$ ./goofys -f --debug_fuse --debug_s3 --profile aws --region s3-ap-northeast-1 --endpoint https://s3-ap-northeast-1.amazonaws.com mybucket /path/to/mount
{code}

**Comment by t2y** [comment]
Also I run goofys with recent sdk version (#402), then it seems the same behavior.

> User-Agent: aws-sdk-go/1.17.13 (go1.12; darwin; amd64)

**Comment by t2y** [comment]
I created a test bucket in `s3-ap-southeast-1` with default configurations on s3 management console. It seems that is the same behavior.

> $ ./goofys -f --debug_fuse --debug_s3 --profile aws --region s3-ap-southeast-1 --endpoint https://s3-ap-southeast-1.amazonaws.com mybucket /path/to/mount

**Comment by t2y** [comment]
I found my faults. I specified aws region name as below.

```
--region s3-ap-northeast-1
--region s3-ap-southeast-1
```

These are wrong names, correct names are as below.

```
--region ap-northeast-1
--region ap-southeast-1
```

Those make the head-object response status code 400 instead of 404! The `testBucket()` expects the response 404 with non-existent object key, but 400 was returned with my wrong region name.

* https://github.com/kahing/goofys/blob/27053313e6161d527f7f7efb1499a46fee1872e7/internal/goofys.go#L276

So, goofys uses v4 signature by default, then uses v2 signature as fallback.
